core/src/main/java/io/questdb/cutlass/line/tcp/DelegatingTlsChannel.java

```diff
@@ -0,0 +1,435 @@
+    }
+
+    @Override
+    public int receive(long ptr, int len) {
+        try {
+            unwrapLoop();
+            unwrapOutputBuffer.flip();
+            int i = unwrapOutputBufferToPtr(ptr, len);
+            unwrapOutputBuffer.compact();
+            return i;
+        } catch (SSLException e) {
+            throw new LineSenderException("error while receiving data from questdb server", e);
+        }
+    }
+
+    private int unwrapOutputBufferToPtr(long dstPtr, int dstLen) {
+        int oldPosition = unwrapOutputBuffer.position();
+
+        assert Unsafe.getUnsafe().getLong(unwrapOutputBufferPtr, ADDRESS_FIELD_OFFSET) == unwrapOutputBufferPtr;
+        long srcPtr = unwrapOutputBufferPtr + oldPosition;
+        int srcLen = unwrapOutputBuffer.remaining();
+        int len = Math.min(dstLen, srcLen);
+        Vect.memcpy(dstPtr, srcPtr, len);
+        unwrapOutputBuffer.position(oldPosition + len);
+        return len;
+    }
+
+    private void growWrapOutputBuffer() {
+        wrapOutputBufferPtr = expandBuffer(wrapOutputBuffer, wrapOutputBufferPtr);
+    }
+
+    private void growUnwrapOutputBuffer() {
+        unwrapOutputBufferPtr = expandBuffer(unwrapOutputBuffer, unwrapOutputBufferPtr);
+    }
+
+    private void growUnwrapInputBuffer() {
+        unwrapInputBufferPtr = expandBuffer(unwrapInputBuffer, unwrapInputBufferPtr);
+    }
+
+    private static long expandBuffer(ByteBuffer buffer, long oldAddress) {
+        int oldCapacity = buffer.capacity();
+        int newCapacity = oldCapacity * 2;
+        long newAddress = Unsafe.realloc(oldAddress, oldCapacity, newCapacity, MemoryTag.NATIVE_DEFAULT);
+        resetBufferToPointer(buffer, newAddress, newCapacity);
+        return newAddress;
+    }
+
+    private static long allocateMemoryAndResetBuffer(ByteBuffer buffer, int capacity) {
+        long newAddress = Unsafe.malloc(capacity, MemoryTag.NATIVE_DEFAULT);
+        resetBufferToPointer(buffer, newAddress, capacity);
+        return newAddress;
+    }
+
+    private static void resetBufferToPointer(ByteBuffer buffer, long ptr, int len) {
+        assert buffer.isDirect();
+        Unsafe.getUnsafe().putLong(buffer, ADDRESS_FIELD_OFFSET, ptr);
+        Unsafe.getUnsafe().putLong(buffer, LIMIT_FIELD_OFFSET, len);
+        Unsafe.getUnsafe().putLong(buffer, CAPACITY_FIELD_OFFSET, len);
+    }
+
+    @Override
+    public int errno() {
+        return delegate.errno();
+    }
+
+    public void close0(boolean closeDelegate) throws IOException {
+        int prevState = state;
+        state = CLOSING;
+        if (prevState == AFTER_HANDSHAKE) {
+            try {
+                sslEngine.closeOutbound();
+                wrapLoop(dummyBuffer);
+                writeToUpstreamAndClear();
+            } catch (Throwable e) {
+                // best effort TLS close signal
+            }
+        }
+        state = CLOSED;
+
+        if (closeDelegate) {
+            try {
+                Misc.free(delegate);
```

Comment:
we're breaking our own convention here. If close() method doesn't throw exceptions, why are we catching them?